shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -72,13 +81,102 @@
     private UUID writerIdPerformingAppends;
     private int currentBlockSize;
     private int bytesLeftInBlock;
+    private final Map<UUID, Session> pendingWrites = new HashMap<>();
 
     @RequiredArgsConstructor
-    private static final class Session {
+    private final class Session {
+        private static final int MAX_EVENTS = 500;
+        private static final int MAX_DATA_SIZE = 1024 * 1024; // 1MB
         private final UUID id;
-        private long lastEventNumber = -1L;
-        private int eventCount;
         private final long requestId;
+        private final List<ByteBuf> pendingList = new ArrayList<>();
+        private int  pendingBytes = 0;
+        private long lastEventNumber = -1L;
+        private int eventCount = 0;
+
+        /**
+         * Record the given Append to Session by tracking the last event number and number of events to write.
+         */
+        private void record(Append append) {
+            lastEventNumber = append.getEventNumber();
+            eventCount += append.getEventCount();
+        }
+
+        /**
+         * Check if session is empty or not.
+         * @return true if there are any pending events; false otherwise.
+         */
+        private boolean isFree() {
+            return eventCount == 0;
+        }
+
+
+        /**
+         * queue the Append data to Session' list.
+         * @param data  data bytes.
+         * @param out   Network channel buffer.
+         */
+        private void write(ByteBuf data, ByteBuf out) {
+            pendingWrites.putIfAbsent(id, this);
```

Comment:
What if this is already present (with this instance or another one)?

Reply:
it will always one instance of session ; when the flush happens its removed; otherwise on write its checked if its already in the list or not; if not add the pendingWrites